src/field_theory/primitive_element.lean

```diff
@@ -0,0 +1,327 @@
+    rw set.mem_set_of_eq at hc,
+    tauto, },
+  have s_fin : s.finite,
+  { apply (set.finite.image r (set.finite_mem_finset (sf.to_finset.product sg.to_finset))).subset,
+    simpa only [set.subset_def, set.mem_image, prod.exists, multiset.mem_to_finset, finset.mem_product] using hr },
+  have s'_fin : (ϕ⁻¹' s).finite := s_fin.preimage ((ring_hom.injective ϕ).inj_on (⇑ϕ⁻¹' s)),
+  obtain ⟨c, hc⟩ := infinite.exists_not_mem_finset s'_fin.to_finset,
+  rw [set.finite.mem_to_finset, set.mem_preimage, set.mem_set_of_eq] at hc,
+  push_neg at hc,
+  exact ⟨c, hc⟩,
+end
+
+lemma primitive_element_two_inf_key_aux {β : F} {h : polynomial F} (h_ne_zero : h ≠ 0) (h_sep : h.separable)
+  (h_root : h.eval β = 0) (h_splits : polynomial.splits ϕ h) (h_roots : ∀ x ∈ (h.map ϕ).roots, x = ϕ β) :
+  h = (polynomial.C (polynomial.leading_coeff h)) * (polynomial.X - polynomial.C β) :=
+begin
+  apply polynomial.map_injective _ ϕ.injective,
+  rw polynomial.splits_iff_exists_multiset at h_splits,
+  cases h_splits with s hs,
+  have h' := @multiset.eq_repeat_of_mem _
+               (polynomial.X - polynomial.C (ϕ β))
+               (s.map (λ (a : E), polynomial.X - polynomial.C a))
+               (λ x hx, begin
+    rw multiset.mem_map at hx,
+    cases hx with a ha,
+    rw h_roots a at ha,
+    exact ha.2.symm,
+    rw polynomial.mem_roots (show polynomial.map ϕ h ≠ 0, by exact polynomial.map_ne_zero h_ne_zero),
+    cases multiset.exists_cons_of_mem ha.1 with y hy,
+    simp [hs, hy],
+  end),
+  rw [h', multiset.prod_repeat, multiset.card_map] at hs,
+  have hf : ¬is_unit (polynomial.X - polynomial.C (ϕ β)),
+  { rw [polynomial.is_unit_iff_degree_eq_zero,polynomial.degree_X_sub_C],
+    exact dec_trivial, },
+  have hn : s.card ≠ 0,
+  { intro hs_card,
+    rw [hs_card, pow_zero, mul_one, ←polynomial.map_C] at hs,
+    rw [polynomial.map_injective _ ϕ.injective hs, polynomial.eval_C, polynomial.leading_coeff_eq_zero] at h_root,
+    exact h_ne_zero h_root, },
+  have h_map_separable : (h.map ϕ).separable := polynomial.separable.map h_sep,
+  rw hs at h_map_separable,
+  rw [(polynomial.separable.of_pow hf hn (polynomial.separable.of_mul_right h_map_separable)).2, pow_one] at hs,
+  simp [hs],
+end
+
+end
+
+variables {F : Type*} [field F] {E : Type*} [field E] [algebra F E]
+
+lemma primitive_element_two_inf_key (α β : E) [F_sep : is_separable F E] (F_inf : infinite F) :
+  ∃ c : F, β ∈ F⟮α + (algebra_map F E) c * β⟯ :=
+begin
+  rcases F_sep α with ⟨hα, hf⟩,
+  rcases F_sep β with ⟨hβ, hg⟩,
+  let f := minimal_polynomial hα,
+  let g := minimal_polynomial hβ,
+  let ιFE := algebra_map F E,
+  let g' := g.map ιFE,
+  let E' := polynomial.splitting_field g',
+  let ιEE' := algebra_map E E',
+  let ιFE' := ιEE'.comp(ιFE),
+  have key := primitive_element_two_aux ιFE' (ιEE' α) (ιEE' β) f g,
+  cases key with c hc,
+  use c,
+  let γ := α+(ιFE c)*β,
+  change β ∈ F⟮γ⟯,
+  let f' := (f.map ιFE).comp(polynomial.C γ-(polynomial.C (ιFE c)) * (polynomial.X)),
```

Comment:
Could you add a statement `have : eval₂ _ f' = ... eval₂ _ f` here? It seems like you use this property in a few places.

Reply:
It looks like there are two places where there are rewrites involving evaluating `f'`. One is in `h_root` and uses `polynomial.eval`, and the other is in `h_roots` and uses `polynomial.eval₂`. It might be possible to combine these, but I'm not sure if it's worth it.